modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/DefaultRuleSet.kt

```diff
@@ -16,15 +16,16 @@ public class DefaultRuleSet : RuleSet by ReflectiveRuleSet(
 	FailFastActionsRule::class,
 	IdNamingRule::class,
 	ImplicitStatusCheckRule::class,
+	InvalidExpressionUsageRule::class,
 	JobDependenciesRule::class,
 	MissingGhRepoRule::class,
 	MissingGhTokenRule::class,
 	MissingJobTimeoutRule::class,
 	MissingNameRule::class,
 	MissingShellRule::class,
-	InvalidExpressionUsageRule::class,
 	PreferGitHubTokenRule::class,
 	RedundantShellRule::class,
+	RequiredPermissionsRule::class,
```

Comment:
Soo, I spent a bit of time writing more tests and thinking a bit more on how this will scale to more actions' permissions. My conclusion: the testing is very complicated because the code path combinations to cover are enormous, it has to cover:
 * effective permissions (tested separately)
 * job vs workflow level permissions (tested separately, but could have more as part of algo)
 * permission inference (we're high-level integration testing each lambda, which seems bad)
   * input token existence
   * github token parsing (testable by increasing visibility)
   * other inputs
 * missing permissions algorithm
 * report formatting

I propose the following:

1. Rebrand this PR to "Model basic permissions"
   1. Delete RequiredPermissionsRule.kt and its tests from this PR
   2. The rest is LGTM
   3. Merge
2. Create another PR where we implement the actual rule (285)
   1. Starting by reverting the deleted 2 files
   2. Then applying something like https://github.com/TWiStErRob/net.twisterrob.ghlint/commit/53dcf9536605efb9a4bd576b799e773487b68619 (design / naming TBD)  
      goal: testable inference and detection algorithm separately from rule)
   3. Doing the same for other actions and their permissions
   4. Testing the rule's core with an injected fake action where the test controls the inference logic

As an optional 1.5 step, we could:
 1. Create a PR which fixes 285
 2. Convert the rest of the commented inference to code (almost without test)
 3. Review and merge
 4. Then follow up with the above proposed "explode" refactor

1.5 is "make it work" and 2 is "make it nice"

Reply:
Agree - but inverted slightly:

https://github.com/TWiStErRob/net.twisterrob.ghlint/pull/318

Mainly because I think the `effective` stuff is quite tied to the rule implementation. 

Once we can get that merged, and then merge `main` back into this pull request, this PR should be in a good place size-wise to play around with a bit more.

Reply:
The idea was to close off this PR and start clean (less commits, less conversation) 😅
But :shrug: at least the amount of lines changed is decreasing.

Thoughts on 1.5 vs 2?

Reply:
I think 2 - but maybe let's keep the number of sub-rules to perhaps 2 to start with? So we can get the improvement live.

Reply:
Ok, here's I think a priority order of the rules we know of so far:
 * checkout
 * sarif
 * deploy-pages
 * encricomi
 * stale
 * eightbit

Reply:
New candidate: dorny/paths-filter, probably slots in above stale.